sharding/handler/shard_tracker.py

```diff
@@ -1,107 +1,136 @@
+            event_name='ReleaseNotary',
+            notary=to_log_topic_address(notary)
+        )
+        return False if not log else True
+
+    def is_new_header_added(self, period):
+        # Get the header added in the specified period
+        log = self._get_logs_by_shard_id(
+            event_name='AddHeader',
+            from_block=period * self.config['PERIOD_LENGTH'],
+            to_block=(period + 1) * self.config['PERIOD_LENGTH'] - 1,
         )
-        for log in new_logs:
-            yield parse_collation_added_log(log)
-
-    def get_next_log(self):
-        new_logs = self._get_new_logs()
-        self.new_logs.extend(new_logs)
-        if len(self.new_logs) == 0:
-            raise NextLogUnavailable("No more next logs")
-        return self.new_logs.pop()
-
-    # TODO: this method may return wrong result when new logs arrive before the logs inside
-    #       `self.new_logs` are consumed entirely. This issue can be resolved by saving the
-    #       status of `new_logs`, `unchecked_logs`, and `current_score`, when it start to run
-    #       `GUESS_HEAD`. If there is a new block arriving, just restore them to the saved status,
-    #       append new logs to `new_logs`, and re-run `GUESS_HEAD`
-    def fetch_candidate_head(self):
-        # Try to return a log that has the score that we are checking for,
-        # checking in order of oldest to most recent.
-        unchecked_logs = pipe(
-            self.unchecked_logs,
-            enumerate,
-            tuple,
-            reversed,
-            tuple,
+        return False if not log else True
+
+    def has_enough_vote(self, period):
+        # Get the votes submitted in the specified period
+        logs = self._get_logs_by_shard_id(
+            event_name='SubmitVote',
+            from_block=period * self.config['PERIOD_LENGTH'],
+            to_block=(period + 1) * self.config['PERIOD_LENGTH'] - 1,
         )
-        current_score = self.current_score
-
-        for idx, log_entry in unchecked_logs:
-            if log_entry['score'] == current_score:
-                return self.unchecked_logs.pop(idx)
-        # If no further recorded but unchecked logs exist, go to the next
-        # is_new_head = true log
-        while True:
-            try:
-                log_entry = self.get_next_log()
-            # TODO: currently just raise when there is no log anymore
-            except NextLogUnavailable:
-                # TODO: should returns the genesis collation instead or just leave it?
-                raise NoCandidateHead("No candidate head available")
-            if log_entry['is_new_head']:
-                break
-            self.unchecked_logs.append(log_entry)
-        self.current_score = log_entry['score']
-        return log_entry
-
-    def clean_logs(self):
-        self.new_logs = []
-        self.unchecked_logs = []
+        return False if not logs else len(logs) >= self.config['QUORUM_SIZE']
```

Comment:
1. If `LogHandler.get_logs()` only returns list, then `ShardTracker._get_logs_by_notary(..)` only returns list. So I think you can write:
```python
return len(logs) >= self.config['QUORUM_SIZE']
```
2. But, I think we need some protections in `LogHandler.get_logs()` to handle timeout, RPC failure, any error messages from web3 APIs ... and we need to add more error handling in different layers.